File: core/src/main/java/org/rpgwizard/common/assets/TileSet.java
/**
 * Copyright (c) 2015, rpgwizard.org, some files forked from rpgtoolkit.net <info@rpgwizard.org>
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/.
 */
package org.rpgwizard.common.assets;

import java.awt.image.BufferedImage;
import java.util.ArrayList;
import java.util.LinkedList;
import java.util.List;

/**
 * This class is responsible for managing a tilset inside the editor It stores all of the tiles in the set in a big
 * Array of tiles!
 *
 * @author Geoff Wilson
 * @author Joshua Michael Daly
 */
public class TileSet extends AbstractAsset {

    private String name;

    private int tileWidth;
    private int tileHeight;

    private String image;
    private BufferedImage bufferedImage;

    private LinkedList<Tile> tiles;

    public TileSet() {
        super(null);
        image = null;
        tiles = new LinkedList<>();
    }

    /**
     * Creates a new TileSet with a descriptor.
     *
     * @param descriptor
     * @param tileWidth
     * @param tileHeight
     */
    public TileSet(AssetDescriptor descriptor, int tileWidth, int tileHeight) {
        super(descriptor);
        this.tileWidth = tileWidth;
        this.tileHeight = tileHeight;
        image = null;
        tiles = new LinkedList<>();
    }

    public String getName() {
        return name;
    }

    public void setName(String name) {
        this.name = name;
    }

    /**
     * Gets a tile from a specified location in the array.
     *
     * @param index
     *            Index of the array to get the tile from
     * @return Tile object representing the tile from the requested index
     */
    public Tile getTile(int index) {
        return tiles.get(index);
    }

    public int getTileIndex(Tile tile) {
        return tiles.indexOf(tile);
    }

    /**
     * Returns an array of all the tiles in the tiles
     *
     * @return Object array of all the tiles in the tiles
     */
    public LinkedList<Tile> getTiles() {
        return tiles;
    }

    public int getTileWidth() {
        return tileWidth;
    }

    public void setTileWidth(int tileWidth) {
        this.tileWidth = tileWidth;
    }

    public int getTileHeight() {
        return tileHeight;
    }

    public void setTileHeight(int tileHeight) {
        this.tileHeight = tileHeight;
    }

    public String getImage() {
        return image;
    }

    public void setImage(String image) {
        this.image = image;
    }

    public BufferedImage getBufferedImage() {
        return bufferedImage;
    }

    public void setBufferedImage(BufferedImage bufferedImage) {
        this.bufferedImage = bufferedImage;
    }

    public void setTiles(LinkedList<Tile> tiles) {
        this.tiles = tiles;
    }

    /**
     * Adds a new tile to the tiles, it will add the tile at the end of the array
     *
     * @param newTile
     *            Tile object to add to the array
     */
    public void addTile(Tile newTile) {
        tiles.add(newTile);
    }

}

File: engine/engine-ts/src/formats/tileset.js
/*
 * Copyright (c) 2017, rpgwizard.org, some files forked from rpgtoolkit.net <info@rpgwizard.org>
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/.
 */
/* global PATH_BITMAP, rpgwizard, Crafty */

function TileSet(filename) {
    if (rpgwizard.debugEnabled) {
        console.debug("Creating Tileset filename=[%s]", filename);
    }
    this.filename = filename;
}

TileSet.prototype.load = async function () {
    if (rpgwizard.debugEnabled) {
        console.debug("Loading Tileset filename=[%s]", this.filename);
    }

    let response = await fetch(this.filename);
    response = await response.json();
    for (var property in response) {
        this[property] = response[property];
    }

    if (rpgwizard.debugEnabled) {
        console.debug("Finished loading Tileset filename=[%s]", this.filename);
    }

    return this;
};

TileSet.prototype.setReady = function () {
    if (rpgwizard.debugEnabled) {
        console.debug("Setting ready TileSet name=[%s]", this.name);
    }
    
    this.img = Crafty.assets[Crafty.__paths.images + this.image];
    this.tileRows = Math.floor(this.img.height / this.tileHeight);
    this.tileColumns = Math.floor(this.img.width / this.tileWidth);
    this.count = this.tileRows * this.tileColumns;
    
    this._prepareTiles();
};

TileSet.prototype._prepareTiles = function () {
    const canvas = document.createElement("canvas");
    canvas.width = this.img.width;
    canvas.height = this.img.height;
    const ctx = canvas.getContext("2d");
    ctx.drawImage(this.img, 0, 0);
    
    this.tiles = [];
    for (var i = 0; i < this.count; i++) {
        // Converted 1D index to 2D cooridnates.
        var x = i % this.tileColumns;
        var y = Math.floor(i / this.tileColumns);
        var tile = ctx.getImageData(
                x * this.tileWidth, y * this.tileHeight,
                this.tileWidth, this.tileHeight);
        this.tiles.push(tile);
    }
};

TileSet.prototype.getTile = function (index) {
    return this.tiles[index];
};